docs/proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,106 @@
+
+The volumeController uses the cloudprovider to create, delete, attach and detach volumes to nodes. For eg, the logic for provisioning, attaching, and detaching a EBS volume resides in the AWS cloudprovider. The volumeController uses this code to perform its operations.
+
+The routeController configures routes for hosts in Google Cloud. 
+
+The serviceController maintains a list of currently active nodes, and is responsible for creating and deleting LoadBalancers in the underlying cloud. 
+
+##### Cloud Provider in Kubelet
+Moving on to the kubelet, the following cloud provider dependencies exist in kubelet.
+
+ - Find the cloud nodename of the host that kubelet is running on for two reasons :
+      1. To obtain the config map for the kubelet, if one already exists
+      2. To uniquely identify current node using nodeInformer
+      3. To instantiate a reference to the current node object
+ - Find the InstanceID, ProviderID, ExternalID, Zone Info of the node object while initailizing it
+ - Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it 
+ - In case of Google cloud, it allows the provider to configure routes for the node
+ - It allows the cloud provider to post process DNS settings
+ 
+##### Cloud Provider in Kube-apiserver
+ 
+Finally, in the kube-apiserver, the cloud provider is used for trasferring SSH keys to all of the nodes, and within an admission controller for setting labels on persistent volumes.
+ 
+### 2. Startegy for refactoring Kube-Controller-Manager 
+
+In order to create a 100% cloud independent controller manager, the idea is to split the controller-manager into two separate binaries. 
+
+1. Cloud dependent controller-manager binary
+2. Cloud independent controller-manager binary
+
+The cloud dependent binary will run those loops that rely on cloudprovider. The rest of the controllers will be run in the cloud independent controller manager. The decision to run entire controller loops, rather than only the very minute parts that rely on cloud provider makes implementation simple. If drilled down further into the controller loops, it would be significantly harder to disentangle, due to tight coupling of the cloud provider specific code with the rest of the data structures/functions within a controller.
+
+There are four controllers that rely on cloud provider specific code. These are node controller, service controller, route controller and attach detach controller. Copies of each of these controllers have been bundled them together into one binary. This binary registers itself as a controller, and runs the cloud specific controller loops with the user-agent named "external-controller-manager".
+
+RouteController and serviceController are entirely cloud specific. Therefore, it is really simple to run these two controller loops separately. 
```

Comment:
Suggest you factor out getNodeConditionPredicate from serviceController into a core library which is can link in, so that if we change node conditions (quite possible) that all serviceControllers will pick up that change.